File: spark/src/main/scala/org/apache/spark/sql/delta/coordinatedcommits/AbstractBatchBackfillingCommitCoordinatorClient.scala
/*
 * Copyright (2021) The Delta Lake Project Authors.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.apache.spark.sql.delta.coordinatedcommits

import java.nio.file.FileAlreadyExistsException
import java.util.{Optional, UUID}

import scala.collection.JavaConverters._

import org.apache.spark.sql.delta.DeltaLog
import org.apache.spark.sql.delta.TransactionExecutionObserver
import org.apache.spark.sql.delta.actions.CommitInfo
import org.apache.spark.sql.delta.actions.Metadata
import org.apache.spark.sql.delta.logging.DeltaLogKeys
import org.apache.spark.sql.delta.util.FileNames
import io.delta.storage.LogStore
import io.delta.storage.commit.{CommitCoordinatorClient, CommitFailedException => JCommitFailedException, CommitResponse, TableDescriptor, TableIdentifier, UpdatedActions}
import org.apache.hadoop.conf.Configuration
import org.apache.hadoop.fs.{FileStatus, Path}

import org.apache.spark.internal.{LoggingShims, MDC}

/**
 * An abstract [[CommitCoordinatorClient]] which triggers backfills every n commits.
 * - every commit version which satisfies `commitVersion % batchSize == 0` will trigger a backfill.
 */
trait AbstractBatchBackfillingCommitCoordinatorClient
  extends CommitCoordinatorClient
    with LoggingShims {

  /**
   * Size of batch that should be backfilled. So every commit version which satisfies
   * `commitVersion % batchSize == 0` will trigger a backfill.
   */
  val batchSize: Long

  /**
   * Commit a given `commitFile` to the table represented by given `logPath` at the
   * given `commitVersion`
   */
  private[delta] def commitImpl(
      logStore: LogStore,
      hadoopConf: Configuration,
      logPath: Path,
      coordinatedCommitsTableConf: Map[String, String],
      commitVersion: Long,
      commitFile: FileStatus,
      commitTimestamp: Long): CommitResponse

  override def commit(
      logStore: LogStore,
      hadoopConf: Configuration,
      tableDesc: TableDescriptor,
      commitVersion: Long,
      actions: java.util.Iterator[String],
      updatedActions: UpdatedActions): CommitResponse = {
    val logPath = tableDesc.getLogPath
    val executionObserver = TransactionExecutionObserver.getObserver
    val tablePath = CoordinatedCommitsUtils.getTablePath(logPath)
    if (commitVersion == 0) {
      throw new JCommitFailedException(false, false, "Commit version 0 must go via filesystem.")
    }
    logInfo(log"Attempting to commit version " +
      log"${MDC(DeltaLogKeys.VERSION, commitVersion)} on table " +
      log"${MDC(DeltaLogKeys.PATH, tablePath)}")
    val fs = logPath.getFileSystem(hadoopConf)
    if (batchSize <= 1) {
      // Backfill until `commitVersion - 1`
      logInfo(log"Making sure commits are backfilled until " +
        log"${MDC(DeltaLogKeys.VERSION, commitVersion - 1)} version for" +
        log" table ${MDC(DeltaLogKeys.PATH, tablePath.toString)}")
      backfillToVersion(
        logStore,
        hadoopConf,
        tableDesc,
        commitVersion - 1,
        null)
    }

    // Write new commit file in _commits directory
    val fileStatus = CoordinatedCommitsUtils.writeCommitFile(
      logStore, hadoopConf, logPath, commitVersion, actions.asScala, generateUUID())

    // Do the actual commit
    val commitTimestamp = updatedActions.getCommitInfo.getCommitTimestamp
    var commitResponse =
      commitImpl(
        logStore,
        hadoopConf,
        logPath,
        tableDesc.getTableConf.asScala.toMap,
        commitVersion,
        fileStatus,
        commitTimestamp)

    val mcToFsConversion = isCoordinatedCommitsToFSConversion(commitVersion, updatedActions)
    // Backfill if needed
    executionObserver.beginBackfill()
    if (batchSize <= 1) {
      // Always backfill when batch size is configured as 1
      backfill(logStore, hadoopConf, logPath, commitVersion, fileStatus)
      val targetFile = FileNames.unsafeDeltaFile(logPath, commitVersion)
      val targetFileStatus = fs.getFileStatus(targetFile)
      val newCommit = commitResponse.getCommit.withFileStatus(targetFileStatus)
      commitResponse = new CommitResponse(newCommit)
    } else if (commitVersion % batchSize == 0 || mcToFsConversion) {
      logInfo(log"Making sure commits are backfilled till " +
        log"${MDC(DeltaLogKeys.VERSION, commitVersion)} " +
        log"version for table ${MDC(DeltaLogKeys.PATH, tablePath.toString)}")
      backfillToVersion(logStore, hadoopConf, tableDesc, commitVersion, null)
    }
    logInfo(log"Commit ${MDC(DeltaLogKeys.VERSION, commitVersion)} done successfully on table " +
      log"${MDC(DeltaLogKeys.PATH, tablePath)}")
    commitResponse
  }

  private def isCoordinatedCommitsToFSConversion(
      commitVersion: Long,
      updatedActions: UpdatedActions): Boolean = {
    val oldMetadataHasCoordinatedCommits =
      CoordinatedCommitsUtils.getCommitCoordinatorName(updatedActions.getOldMetadata).nonEmpty
    val newMetadataHasCoordinatedCommits =
      CoordinatedCommitsUtils.getCommitCoordinatorName(updatedActions.getNewMetadata).nonEmpty
    oldMetadataHasCoordinatedCommits && !newMetadataHasCoordinatedCommits && commitVersion > 0
  }

  protected def generateUUID(): String = UUID.randomUUID().toString

  override def backfillToVersion(
      logStore: LogStore,
      hadoopConf: Configuration,
      tableDesc: TableDescriptor,
      version: Long,
      lastKnownBackfilledVersionOpt: java.lang.Long): Unit = {
    val logPath = tableDesc.getLogPath
    // Confirm the last backfilled version by checking the backfilled delta file's existence.
    val validLastKnownBackfilledVersionOpt = Option(lastKnownBackfilledVersionOpt)
        .filter { version =>
      val fs = logPath.getFileSystem(hadoopConf)
      fs.exists(FileNames.unsafeDeltaFile(logPath, version))
    }
    val startVersionOpt: Long = validLastKnownBackfilledVersionOpt.map(_ + 1).map(Long.box).orNull
    getCommits(tableDesc, startVersionOpt, version)
      .getCommits.asScala
      .foreach { commit =>
        backfill(logStore, hadoopConf, logPath, commit.getVersion, commit.getFileStatus)
    }
  }

  /** Backfills a given `fileStatus` to `version`.json */
  protected def backfill(
      logStore: LogStore,
      hadoopConf: Configuration,
      logPath: Path,
      version: Long,
      fileStatus: FileStatus): Unit = {
    val targetFile = FileNames.unsafeDeltaFile(logPath, version)
    logInfo(log"Backfilling commit ${MDC(DeltaLogKeys.PATH, fileStatus.getPath)} to " +
      log"${MDC(DeltaLogKeys.PATH2, targetFile.toString)}")
    val commitContentIterator = logStore.read(fileStatus.getPath, hadoopConf)
    try {
      logStore.write(
        targetFile,
        commitContentIterator,
        false,
        hadoopConf)
      registerBackfill(logPath, version)
    } catch {
      case _: FileAlreadyExistsException =>
        logInfo(log"The backfilled file ${MDC(DeltaLogKeys.FILE_NAME, targetFile)} already exists.")
    } finally {
      commitContentIterator.close()
    }
  }

  /**
   * Callback to tell the CommitCoordinator that all commits <= `backfilledVersion` are backfilled.
   */
  protected[delta] def registerBackfill(
      logPath: Path,
      backfilledVersion: Long): Unit

}

File: kernel/kernel-defaults/src/main/java/io/delta/kernel/defaults/internal/coordinatedcommits/AbstractBatchBackfillingCommitCoordinatorClient.java
/*
 * Copyright (2023) The Delta Lake Project Authors.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package io.delta.kernel.defaults.internal.coordinatedcommits;

import java.io.IOException;
import java.nio.file.FileAlreadyExistsException;
import java.util.Iterator;
import java.util.Map;
import java.util.UUID;

import io.delta.storage.CloseableIterator;
import io.delta.storage.LogStore;
import io.delta.storage.commit.Commit;
import io.delta.storage.commit.CommitCoordinatorClient;
import io.delta.storage.commit.CommitFailedException;
import io.delta.storage.commit.CommitResponse;
import io.delta.storage.commit.GetCommitsResponse;
import io.delta.storage.commit.UpdatedActions;
import org.apache.hadoop.conf.Configuration;
import org.apache.hadoop.fs.FileStatus;
import org.apache.hadoop.fs.FileSystem;
import org.apache.hadoop.fs.Path;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

/**
 * An abstract {@link CommitCoordinatorClient} which triggers backfills every n commits.
 * - every commit version which satisfies `commitVersion % batchSize == 0` will trigger a backfill.
 */
public abstract class AbstractBatchBackfillingCommitCoordinatorClient
        implements CommitCoordinatorClient {

    protected static final Logger logger =
            LoggerFactory.getLogger(AbstractBatchBackfillingCommitCoordinatorClient.class);

    /**
     * Size of batch that should be backfilled. So every commit version which satisfies
     * `commitVersion % batchSize == 0` will trigger a backfill.
     */
    protected long batchSize;

    /**
     * Commit a given `commitFile` to the table represented by given `logPath` at the
     * given `commitVersion`
     */
    protected abstract CommitResponse commitImpl(
            LogStore logStore,
            Configuration hadoopConf,
            Path logPath,
            Map<String, String> coordinatedCommitsTableConf,
            long commitVersion,
            FileStatus commitFile,
            long commitTimestamp) throws CommitFailedException;

    @Override
    public CommitResponse commit(
            LogStore logStore,
            Configuration hadoopConf,
            Path logPath,
            Map<String, String> coordinatedCommitsTableConf,
            long commitVersion,
            Iterator<String> actions,
            UpdatedActions updatedActions) throws CommitFailedException, IOException {
        Path tablePath = CoordinatedCommitsUtils.getTablePath(logPath);
        if (commitVersion == 0) {
            throw new CommitFailedException(
                    false, false, "Commit version 0 must go via filesystem.");
        }
        logger.info("Attempting to commit version {} on table {}", commitVersion, tablePath);
        FileSystem fs = logPath.getFileSystem(hadoopConf);
        if (batchSize <= 1) {
            // Backfill until `commitVersion - 1`
            logger.info("Making sure commits are backfilled until {}" +
                    " version for table {}", commitVersion - 1, tablePath);
            backfillToVersion(
                    logStore,
                    hadoopConf,
                    logPath,
                    coordinatedCommitsTableConf,
                    commitVersion - 1,
                    null);
        }

        // Write new commit file in _commits directory
        FileStatus fileStatus = CoordinatedCommitsUtils.writeCommitFile(
                logStore, hadoopConf, logPath.toString(), commitVersion, actions, generateUUID());

        // Do the actual commit
        long commitTimestamp = updatedActions.getCommitInfo().getCommitTimestamp();
        CommitResponse commitResponse =
                commitImpl(
                        logStore,
                        hadoopConf,
                        logPath,
                        coordinatedCommitsTableConf,
                        commitVersion,
                        fileStatus,
                        commitTimestamp);

        boolean mcToFsConversion = isCoordinatedCommitsToFSConversion(
                commitVersion, updatedActions);
        // Backfill if needed
        if (batchSize <= 1) {
            // Always backfill when batch size is configured as 1
            backfill(logStore, hadoopConf, logPath, commitVersion, fileStatus);
            Path targetFile = CoordinatedCommitsUtils.getHadoopDeltaFile(logPath, commitVersion);
            FileStatus targetFileStatus = fs.getFileStatus(targetFile);
            Commit newCommit = commitResponse.getCommit().withFileStatus(targetFileStatus);
            return new CommitResponse(newCommit);
        } else if (commitVersion % batchSize == 0 || mcToFsConversion) {
            logger.info("Making sure commits are backfilled till {} version for table {}",
                    commitVersion,
                    tablePath);
            backfillToVersion(
                    logStore,
                    hadoopConf,
                    logPath,
                    coordinatedCommitsTableConf,
                    commitVersion,
                    null);
        }
        logger.info("Commit {} done successfully on table {}", commitVersion, tablePath);
        return commitResponse;
    }

    @Override
    public void backfillToVersion(
            LogStore logStore,
            Configuration hadoopConf,
            Path logPath,
            Map<String, String> coordinatedCommitsTableConf,
            long version,
            Long lastKnownBackfilledVersion) throws IOException {
        // Confirm the last backfilled version by checking the backfilled delta file's existence.
        if (lastKnownBackfilledVersion != null) {
            try {
                FileSystem fs = logPath.getFileSystem(hadoopConf);
                if (!fs.exists(CoordinatedCommitsUtils.getHadoopDeltaFile(logPath, version))) {
                    lastKnownBackfilledVersion = null;
                }
            } catch (IOException e) {
                throw new RuntimeException(e);
            }
        }
        Long startVersion = null;
        if (lastKnownBackfilledVersion != null) {
            startVersion = lastKnownBackfilledVersion + 1;
        }
        GetCommitsResponse commitsResponse =
                getCommits(logPath,coordinatedCommitsTableConf, startVersion, version);
        for (Commit commit : commitsResponse.getCommits()) {
            backfill(logStore, hadoopConf, logPath, commit.getVersion(), commit.getFileStatus());
        }
    }

    protected String generateUUID() {
        return UUID.randomUUID().toString();
    }

    /** Backfills a given `fileStatus` to `version`.json */
    protected void backfill(
            LogStore logStore,
            Configuration hadoopConf,
            Path logPath,
            long version,
            FileStatus fileStatus) throws IOException {
        Path targetFile = CoordinatedCommitsUtils.getHadoopDeltaFile(logPath, version);
        logger.info("Backfilling commit " + fileStatus.getPath() + " to " + targetFile);
        CloseableIterator<String> commitContentIterator = logStore
                .read(fileStatus.getPath(), hadoopConf);
        try {
            logStore.write(
                    targetFile,
                    commitContentIterator,
                    false,
                    hadoopConf);
            registerBackfill(logPath, version);
        } catch (FileAlreadyExistsException e) {
            logger.info("The backfilled file " + targetFile + " already exists.");
        } finally {
            commitContentIterator.close();
        }
    }

    /**
     * Callback to tell the CommitCoordinator that all commits <= `backfilledVersion` are
     * backfilled.
     */
    protected abstract void registerBackfill(Path logPath, long backfilledVersion);

    private boolean isCoordinatedCommitsToFSConversion(
            long commitVersion, UpdatedActions updatedActions) {
        boolean oldMetadataHasCoordinatedCommits =
                CoordinatedCommitsUtils
                        .getCommitCoordinatorName(updatedActions.getOldMetadata()).isPresent();
        boolean newMetadataHasCoordinatedCommits =
                CoordinatedCommitsUtils
                        .getCommitCoordinatorName(updatedActions.getNewMetadata()).isPresent();
        return oldMetadataHasCoordinatedCommits
                && !newMetadataHasCoordinatedCommits
                && commitVersion > 0;
    }
}
